webExtensionApis/study/src/index.js

```diff
@@ -0,0 +1,395 @@
+         *
+         *  Throws ExtensionError if called before `browser.study.setup`
+         **/
+        getStudyInfo: async function getStudyInfo() {
+          logger.log("called getStudyInfo ");
+          return studyUtils.info();
+        },
+
+        /** Send Telemetry using appropriate shield or pioneer methods.
+         *
+         *  shield:
+         *  - `shield-study-addon` ping, requires object string keys and string values
+         *
+         *  pioneer:
+         *  - TBD
+         *
+         *  Note:
+         *  - no conversions / coercion of data happens.
+         *
+         *  Note:
+         *  - undefined what happens if validation fails
+         *  - undefined what happens when you try to send 'shield' from 'pioneer'
+         *
+         *  TBD fix the parameters here.
+         *
+         * @param {Object} payload Non-nested object with key strings, and key values
+         * @returns {undefined}
+         */
+        sendTelemetry: async function sendTelemetry(payload) {
+          logger.log("called sendTelemetry payload");
+
+          function throwIfInvalid(obj) {
+            // Check: all keys and values must be strings,
+            for (const k in obj) {
+              if (typeof k !== "string")
+                throw new ExtensionError(`key ${k} not a string`);
+              if (typeof obj[k] !== "string")
+                throw new ExtensionError(`value ${k} ${obj[k]} not a string`);
+            }
+            return true;
+          }
+
+          throwIfInvalid(payload);
+          return studyUtils.telemetry(payload);
+        },
+
+        /** Search locally stored telemetry pings using these fields (if set)
+         *
+         *  n:
+         *    if set, no more than `n` pings.
+         *  type:
+         *    Array of 'ping types' (e.g., main, crash, shield-study-addon) to filter
+         *  mininumTimestamp:
+         *    only pings after this timestamp.
+         *  headersOnly:
+         *    boolean.  If true, only the 'headers' will be returned.
+         *
+         *  Pings will be returned sorted by timestamp with most recent first.
+         *
+         *  Usage scenarios:
+         *  - enrollment / eligiblity using recent Telemetry behaviours or client environment
+         *  - addon testing scenarios
+         *
+         * @param {Object<query>} searchTelemetryQuery see above
+         * @returns {Array<sendTelemetry>} matchingPings
+         */
+        async searchSentTelemetry(searchTelemetryQuery) {
+          const { TelemetryArchive } = ChromeUtils.import(
+            "resource://gre/modules/TelemetryArchive.jsm",
+            {},
+          );
+          const { searchTelemetryArchive } = require("./telemetry.js");
+          return searchTelemetryArchive(TelemetryArchive, searchTelemetryQuery);
+        },
+
+        /* Using AJV, do jsonschema validation of an object.  Can be used to validate your arguments, packets at client. */
```

Comment:
Several things:

Firefox includes "Schemas.jsm" 
- url: https://dxr.mozilla.org/mozilla-central/source/toolkit/components/extensions/Schemas.jsm

1. Minor: Undocumented.   unknown set of tests
2. Minor:Not obvious how to use it for general validation.
3. Supports modified / non-standard Draft-03 of jsonschema.  Today is Draft-07.  Draft-05 is pretty back-breaking
4. I want to validation *outside* of firefox during build steps (because it's faster to build with node than fx).  
5.  Ajv is pretty awesomely well supported and tested 3rd party software,.  But it's big.
6.  When we *started* there wasn't any Jsonschema in fx :)

Reply:
(I am very happy to drop this Very Soon.   Jsonschema as a "guard" has been useful.  So would typescript.  Meh :) )